File: backend/app/schemas/syllabus.py
"""Syllabus tracking schemas."""

from datetime import datetime
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class TopicProgressUpdate(BaseModel):
    is_completed: bool


class TopicResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    title: str
    sort_order: int
    is_completed: bool = False
    completed_at: Optional[datetime] = None


class ChapterResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    title: str
    sort_order: int
    topic_count: int
    completed_topic_count: int
    completion_percentage: int
    topics: list[TopicResponse] = Field(default_factory=list)


class SubjectResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    name: str
    code: str
    sort_order: int
    chapter_count: int
    topic_count: int
    completed_topic_count: int
    completion_percentage: int


class SubjectDetailResponse(SubjectResponse):
    chapters: list[ChapterResponse] = Field(default_factory=list)


class SyllabusStructureResponse(BaseModel):
    subjects: list[SubjectDetailResponse]
    overall_completion_percentage: int
    total_topics: int
    completed_topics: int


class SyllabusCompletionResponse(BaseModel):
    overall_completion_percentage: int
    total_subjects: int
    total_chapters: int
    total_topics: int
    completed_topics: int
    subjects: list[SubjectResponse]

File: frontend/types/syllabus.ts
export type Topic = {
  id: string;
  title: string;
  sort_order: number;
  is_completed: boolean;
  completed_at: string | null;
  is_locked: boolean;
  is_current: boolean;
};

export type Chapter = {
  id: string;
  title: string;
  sort_order: number;
  topic_count: number;
  completed_topic_count: number;
  completion_percentage: number;
  topics: Topic[];
};

export type Subject = {
  id: string;
  name: string;
  code: string;
  sort_order: number;
  chapter_count: number;
  topic_count: number;
  completed_topic_count: number;
  completion_percentage: number;
};

export type SubjectDetail = Subject & {
  chapters: Chapter[];
};

export type SyllabusCompletion = {
  overall_completion_percentage: number;
  total_subjects: number;
  total_chapters: number;
  total_topics: number;
  completed_topics: number;
  subjects: Subject[];
};

export type SyllabusStructure = {
  subjects: SubjectDetail[];
  overall_completion_percentage: number;
  total_topics: number;
  completed_topics: number;
};
